### PR TITLE
fix(awful: placement: no_offsceen): use new placement infrastructure

### PR DESCRIPTION
**Pull request by actionless**
first time awesome was crashing during the travis, hope this time it should be fine

fixes https://github.com/awesomeWM/awesome/issues/1092

### CODE REVIEW

File: lib/awful/placement.lua [hunk 1]
```diff
@@ -827,17 +827,19 @@ end
...
+-- @tparam[opt={}] table args The arguments
+-- @tparam[opt=client's screen] integer args.screen The screen.
 -- @treturn table The new client geometry.
-function placement.no_offscreen(c, screen)
+function placement.no_offscreen(c, args)
     --HACK necessary for composition to work. The API will be changed soon
-    if type(screen) == "table" then
-        screen = nil
+    if type(args) == "number" or type(args) == "screen" then
+        args = { screen = args }
```
Comment (actionless) [inline lib/awful/placement.lua]: mb print deprecation warning here

**Comment by actionless** [comment]
weird, now geometry is 4 px bigger in each dimension in those 2 tests

trying to figure out why

**Comment by actionless** [comment]
squahed to fix PREV_COMMITS test

File: lib/awful/placement.lua [hunk 2]
```diff
@@ -827,17 +834,19 @@ end
 --- Place the client so no part of it will be outside the screen (workarea).
 --@DOC_awful_placement_no_offscreen_EXAMPLE@
 -- @client c The client.
--- @tparam[opt=client's screen] integer screen The screen.
+-- @tparam[opt={}] table args The arguments
+-- @tparam[opt=client's screen] integer args.screen The screen.
 -- @treturn table The new client geometry.
-function placement.no_offscreen(c, screen)
+function placement.no_offscreen(c, args)
     --HACK necessary for composition to work. The API will be changed soon
```
Comment (psychon) [inline lib/awful/placement.lua]: This comment is now wrong / outdated, right?

Comment (actionless) [reply lib/awful/placement.lua]: updated the comment and added deprecation warning

**Comment by actionless** [comment]
i am not sure if that error is related to the change:
```
== Running /home/travis/build/awesomeWM/awesome/tests/test-awful-client.lua ==
2018-07-25 12:37:46 W: awesome: a_glib_poll:431: Last main loop iteration took 0.185613 seconds! Increasing limit for this warning to that value.
xterm: cannot load font '-misc-fixed-medium-r-semicondensed--13-120-75-75-c-60-iso10646-1'
xterm: cannot load font '-misc-fixed-medium-r-semicondensed--13-120-75-75-c-60-iso10646-1'
2018-07-25 12:37:47 W: awesome: a_glib_poll:431: Last main loop iteration took 0.346065 seconds! Increasing limit for this warning to that value.
awesome: lvm.c:436: luaV_equalobj: Assertion `(((t1)->value_).gc)->tt == 5' failed.
2018-07-25 12:38:00 E: awesome: signal_fatal:448: signal 6, dumping backtrace
/home/travis/build/awesomeWM/awesome/build/awesome(backtrace_get+0x49) [0x42d722]
/home/travis/build/awesomeWM/awesome/build/awesome() [0x40f2a5]
/lib/x86_64-linux-gnu/libc.so.6(+0x36cb0) [0x2adeb66e5cb0]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0x37) [0x2adeb66e5c37]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x148) [0x2adeb66e9028]
/lib/x86_64-linux-gnu/libc.so.6(+0x2fbf6) [0x2adeb66debf6]
/lib/x86_64-linux-gnu/libc.so.6(+0x2fca2) [0x2adeb66deca2]
/usr/lib/liblua.so(+0x20553) [0x2adeb6244553]
/usr/lib/liblua.so(+0x1cb2d) [0x2adeb6240b2d]
/usr/lib/liblua.so(+0x1d1be) [0x2adeb62411be]
/usr/lib/liblua.so(+0x230c1) [0x2adeb62470c1]
/usr/lib/liblua.so(+0x115ff) [0x2adeb62355ff]
/usr/lib/liblua.so(+0x11651) [0x2adeb6235651]
/usr/lib/liblua.so(+0x104cf) [0x2adeb62344cf]
/usr/lib/liblua.so(+0x11abd) [0x2adeb6235abd]
/usr/lib/liblua.so(lua_pcallk+0xca) [0x2adeb622f72a]
/home/travis/build/awesomeWM/awesome/build/awesome(signal_object_emit+0x375) [0x431d8b]
/home/travis/build/awesomeWM/awesome/build/awesome(luaA_class_emit_signal+0x15) [0x42fcbd]
/home/travis/build/awesomeWM/awesome/build/awesome(luaA_object_emit_signal+0x56f) [0x432433]
/home/travis/build/awesomeWM/awesome/build/awesome() [0x446efd]
/home/travis/build/awesomeWM/awesome/build/awesome() [0x446fd5]
/usr/lib/liblua.so(+0x11113) [0x2adeb6235113]
/usr/lib/liblua.so(+0x214db) [0x2adeb62454db]
/usr/lib/liblua.so(+0x115ff) [0x2adeb62355ff]
/usr/lib/liblua.so(+0x11651) [0x2adeb6235651]
/usr/lib/liblua.so(+0x104cf) [0x2adeb62344cf]
/usr/lib/liblua.so(+0x11abd) [0x2adeb6235abd]
/usr/lib/liblua.so(lua_pcallk+0xca) [0x2adeb622f72a]
/usr/lib/liblua.so(+0x26c20) [0x2adeb624ac20]
/usr/lib/liblua.so(+0x11113) [0x2adeb6235113]
/usr/lib/liblua.so(+0x214db) [0x2adeb62454db]
/usr/lib/liblua.so(+0x115ff) [0x2adeb62355ff]
Awesome exited with status code 1
```

**Comment by actionless** [comment]
not reproducible after test restart

**Comment by actionless** [comment]
and the third time it passed

**Comment by Elv13** [comment]
Thanks for working on this given that's all my fault. We can't really ignore this issue anymore so it has to be merged. I fear side effects are unavoidable and new `busted` test will be needed in the future along with the existing documentation examples tests. Since `no_offscreen` and `no_overlap` are the 2 most troublesome functions of `awful.placement`. It is in part why I left them alone from the old API style instead of porting them in the first place. The "new" (from 2016) API style is great for most other functions, but less than ideal for these 2. That's a bit "sad" given those are the 2 most important ones.

@actionless Any known regression right now? Otherwise I propose to merge it and see what fails. Just like the Awesome v4.1 maximization and Awesome v3.x `awful.rules` execution order fiascos, the only way forward is to bite the bullet and add regression tests over time as issues are discovered and fixed.

**Comment by actionless** [comment]
@Elv13 let's try then :-)

**Comment by psychon** [comment]
Whoops, I missed the fixup commit. Sorry.

Also:
```
[ 26%] Running tests/examples/awful/placement/no_offscreen.lua (via tests/examples/awful/template.lua, generating build/doc/images/AUTOGEN_awful_placement_no_offscreen.svg)
Expected text from /home/psychon/awesome/tests/examples/awful/placement/no_offscreen.output.txt, but got:
--- /home/psychon/awesome/tests/examples/awful/placement/no_offscreen.output.txt	2018-08-10 17:02:06.516197986 +0200
+++ /tmp/tmp.UFmHEz3Pm4	2018-08-22 12:28:52.469458858 +0200
@@ -1,2 +1,2 @@
 Before:	x=-30, y=-30, width=100, height=100
-After:	x=10, y=10, width=100, height=100
+After:	x=50, y=50, width=20, height=20
```
Git bisect blames this on fec8d6aa8fe257f2d1fbab9a908b1e3a9a98bed3.
Testing by hand confirms:
```
$ git checkout fec8d6aa8fe257f2d1fbab9a908b1e3a9a98bed3^
$ make
[...]
$ grep -r 'After:' build
build/lib/awful/placement.lua:--    After:	x=10, y=10, width=100, height=100
build/doc/libraries/awful.placement.html:After:  x=10, y=10, width=100, height=100
$ git checkout fec8d6aa8fe257f2d1fbab9a908b1e3a9a98bed3^
$ make
[...]
$ grep -r 'After:' build
build/lib/awful/placement.lua:--    After:	x=50, y=50, width=20, height=20
build/doc/libraries/awful.placement.html:After:  x=50, y=50, width=20, height=20
```
Thoughts about this?

**Comment by actionless** [comment]
i am wondering  why it didn't failed on travis, mb it was false positive all the time because of some other bugfixes which were recently pushed to master (after travis was running against this PR)

i think the reason is `tests/examples/awful/placement/no_offscreen.lua`, this PR got uncommented margins=40 there

**Comment by psychon** [comment]
Yeah, I added the `.txt` files with expected output only while this PR was open. Still, can someone explain why previously ` x=10, y=10, width=100, height=100` was the expected result for a client that started with `x=-30, y=-30, width=100, height=100` (which makes some sense to me, but I am not quite sure where that `x=10, y=10` comes from), but suddenly the expected result is `x=50, y=50, width=20, height=20`? Is this just because the `margins=40` were ignored before and are actually honored now?

**Comment by psychon** [comment]
I went ahead and opened #2369

**Comment by actionless** [comment]
thanks!

**Pull request merged**